analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -23,29 +23,47 @@ namespace SonarAnalyzer.CFG.Roslyn
     public abstract class CfgAllPathValidator
     {
         private readonly ControlFlowGraph cfg;
-        private readonly Dictionary<BasicBlock, bool> visitedStatus = new Dictionary<BasicBlock, bool>();
+        private readonly HashSet<BasicBlock> visitedStatus = [];
 
         protected abstract bool IsValid(BasicBlock block);
         protected abstract bool IsInvalid(BasicBlock block);
 
         protected CfgAllPathValidator(ControlFlowGraph cfg) =>
             this.cfg = cfg;
 
-        public bool CheckAllPaths() =>
-            IsBlockOrAllSuccessorsValid(cfg.EntryBlock);
-
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
-        {
-            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
-            visitedStatus[block] = isValid;
-            return isValid;
-        }
-
-        private bool AreAllSuccessorsValid(BasicBlock block)
+        public bool CheckAllPaths()
         {
-            visitedStatus[block] = true; // protects from loops, don't fail the computation if hits itself
-            return block.SuccessorBlocks.Any()
-                   && block.SuccessorBlocks.All(x => x != cfg.ExitBlock && (visitedStatus.ContainsKey(x) ? visitedStatus[x] : IsBlockOrAllSuccessorsValid(x)));
+            var blocks = new Stack<BasicBlock>();
+            blocks.Push(cfg.EntryBlock);
+            while (blocks.Count > 0)
+            {
+                var block = blocks.Pop();
+                if (IsInvalid(block))
+                {
+                    return false;
+                }
+                visitedStatus.Add(block); // protects from loops, don't fail the computation if hits itself
+                if (IsValid(block))
+                {
+                    continue;
+                }
+                if (block.SuccessorBlocks.IsEmpty)
+                {
+                    return false;
+                }
+                foreach (var successorBlock in block.SuccessorBlocks)
+                {
+                    if (successorBlock == cfg.ExitBlock)
+                    {
+                        return false;
+                    }
+                    if (!visitedStatus.Contains(successorBlock))
```

Comment:
I think there's a gap in this check that allows re-visiting the same block multiple times. If we push blocks `A` and `B`, we don't add them to visited. So if block `A` would want to push `B` again, it will. And we will analyze `B` multiple times.

This `if` should be removed, to push duplicates if needed. Instead, we would check visited as the 1st step after poping:
```
var block = blocks.Pop();
if (visited.Add(block)) // protects from loops, don't fail the computation if hits itself
{
  ... the rest of implementation
}
```